src/main/java/org/openjdk/shenandoah/RegionPopUp.java

```diff
@@ -48,52 +41,55 @@ public class RegionPopUp extends JFrame {
 
     Snapshot snapshot;
 
-    public RegionPopUp(int regionx, int regiony, Snapshot snapshot, int regionWidth, int regionHeight) {
-        this.regionx = regionx;
-        this.regiony = regiony;
+    public RegionPopUp(Snapshot snapshot, int regionNumber) {
         this.snapshot = snapshot;
-        this.regionWidth = regionWidth;
-        this.regionHeight = regionHeight;
+        this.regionNumber = regionNumber;
         JPanel detailedState = new JPanel() {
             public void paint(Graphics g) {
                 renderDetailedRegion(g);
             }
         };
         this.add(detailedState);
-        area = regionWidth * regionHeight;
-        sqSize = Math.max(1, (int) Math.sqrt(1D * area / snapshot.regionCount()));
-        cols = regionWidth / sqSize;
-        regionNumber = (((regionx / sqSize) + 1) + ((regiony / sqSize) * cols)) - 1;
 //        System.out.println(regionNumber);
-        if (regionNumber >= 0 && regionNumber < snapshot.statsSize()) {
-            usedLvl = snapshot.get(regionNumber).used() * 100f;
-            liveLvl = snapshot.get(regionNumber).live() * 100f;
-            tlabLvl = snapshot.get(regionNumber).tlabAllocs() * 100f;
-            gclabLvl = snapshot.get(regionNumber).gclabAllocs() * 100f;
-            plabLvl = snapshot.get(regionNumber).plabAllocs() * 100f;
-            sharedLvl = snapshot.get(regionNumber).sharedAllocs() * 100f;
-            state = snapshot.get(regionNumber).state();
-            age = snapshot.get(regionNumber).age();
-            affiliation = snapshot.get(regionNumber).affiliation();
-        }
+        usedLvl = snapshot.get(regionNumber).used() * 100f;
```

Comment:
We can get rid of some code duplication here if we have the constructor call `setSnapshot`. Just to be safe, mark the `setSnapshot` method as `final`. Any method called from the constructor on the class being constructed should be `final`, or you can have weird crashes if virtual methods resolve to a derived class. Note that `private` methods are `final` by default. You could also use a local variable here to avoid repeated look ups, like:

```
regionData = snapshot.get(regionNumber);
liveLvl = regionData.live() * 100f;
```